compiler/src/iree/compiler/Codegen/Common/FoldAffineMinWithSCFIf.cpp

```diff
@@ -0,0 +1,199 @@
+  scf::ForOp forOp = scf::getForInductionVarOwner(iv);
+  if (!forOp) return failure();
+
+  auto loopInfo = isTiledAndDistributedLoop(forOp);
+  if (!loopInfo) return failure();
+
+  Optional<int64_t> untiledStep = getConstantIntValue(loopInfo->untiledStep);
+  // For IREE right now the original untiled loop should have step 1..
+  if (!untiledStep || *untiledStep != 1) return failure();
+  // ..and we tile according to some static tile sizes for processors.
+  if (!loopInfo->tileSize) return failure();
+
+  lb = loopInfo->untiledLowerBound;
+  ub = loopInfo->untiledUpperBound;
+  // The "step" expected by the upstream utility is really the tiling size.
+  step = OpBuilder(iv.getContext()).getIndexAttr(loopInfo->tileSize.value());
+  return success();
+}
+
+/// Match affine.min affine_map<(d0) -> (-d0 + UB, TILESIZE)>(%iv) from a
+/// tile-and-distribution loop.
+static LogicalResult matchBoundedTileSizeOp(AffineMinOp minOp, int64_t &lb,
+                                            int64_t &ub, int64_t &tileSize) {
+  if (minOp.getNumOperands() != 1) return failure();
+
+  AffineMap map = minOp.getAffineMap();
+  if (map.getNumResults() != 2) return failure();
+
+  OpFoldResult lbOfr, ubOfr, tileSizeOfr;
+  if (failed(matchTileAndDistributionLoop(minOp.getOperand(0), lbOfr, ubOfr,
+                                          tileSizeOfr)))
+    return failure();
+
+  Optional<int64_t> lbOpt = getConstantIntValue(lbOfr);
+  if (!lbOpt) return failure();
+  Optional<int64_t> ubOpt = getConstantIntValue(ubOfr);
+  if (!ubOpt) return failure();
+  Optional<int64_t> tileSizeOpt = getConstantIntValue(tileSizeOfr);
+  if (!tileSizeOpt) return failure();
+
+  lb = *lbOpt;
+  ub = *ubOpt;
+  tileSize = *tileSizeOpt;
+  return success();
+}
+
+/// Folds `affine.min` if its constant value argument is used by scf.if,
+/// where the op is folded into one of the two values.
+///
+/// ```mlir
+/// %min = affine.min(%a, %cst)
+/// %cmp = icmp.eq %min, %cst
+/// scf.if %cmp {
+///   // expect the uses of %min are already replaced by %cst
+/// } else {
+///   // use of %min can be replaced by %a
+/// }
+/// ```
+struct FoldAffineMinWithSCFIf final : public OpRewritePattern<scf::IfOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(scf::IfOp ifOp,
+                                PatternRewriter &rewriter) const override {
+    auto cmpOp = ifOp.getCondition().getDefiningOp<arith::CmpIOp>();
+    if (!cmpOp) return failure();
+    if (cmpOp.getPredicate() != arith::CmpIPredicate::eq) return failure();
+
+    AffineMinOp minOp;
+    Value b;
+    if (isa<AffineMinOp>(cmpOp.getLhs().getDefiningOp())) {
+      minOp = cast<AffineMinOp>(cmpOp.getLhs().getDefiningOp());
+      b = cmpOp.getRhs();
+    } else if (isa<AffineMinOp>(cmpOp.getRhs().getDefiningOp())) {
+      minOp = cast<AffineMinOp>(cmpOp.getRhs().getDefiningOp());
+      b = cmpOp.getLhs();
+    } else {
+      return failure();
+    }
```

Comment:
For now, the specialization has only two cases, i.e., specialized or not. It is mainly because of the code duplication for multi dimensional cases, which explodes the code size by 2^(num_dimensions). We would get 4 and 8 cases for 2D and 3D. We can extend it later if specializing all the cases would show a clear benefit at the expense of the code size.